magenta-lib/src/main/scala/magenta/graph/DeploymentGraph.scala

```diff
@@ -0,0 +1,90 @@
+package magenta.graph
+
+import magenta.tasks.Task
+
+sealed trait Node {
+  def ~>(to: Node) = Edge(this, to)
+  def maybePriority: Option[Int] = None
+}
+case class StartNode() extends Node
+case class DeploymentNode(tasks: List[Task], pathName: String, priority: Int = 1) extends Node {
+  override def maybePriority: Option[Int] = Some(priority)
+  def incPriority(n: Int): DeploymentNode = this.copy(priority = priority + n)
+}
+case class EndNode() extends Node
+
+case class Edge(from: Node, to: Node)
+
+/**
+  * This graph is specialised for deployment. It consists of a start node, an end node and any number of
+  * deployment nodes. The graph structure represents which deployments must be run in series and which can be run in
+  * parallel.
+  */
+object DeploymentGraph {
+  def apply(tasks: List[Task], pathName: String, priority: Int = 1): DeploymentGraph = {
+    val deploymentNode = DeploymentNode(tasks, pathName, priority)
+    DeploymentGraph(StartNode() ~> deploymentNode, deploymentNode ~> EndNode())
+  }
+
+  def apply(edges: Edge*): DeploymentGraph = DeploymentGraph(edges.toSet)
+
+  val empty = DeploymentGraph(StartNode() ~> EndNode())
+}
+
+case class DeploymentGraph(edges: Set[Edge]) {
```

Comment:
I can't help feeling that a lot of this would be simpler if our view was oriented around nodes, rather than edges.